### PR TITLE
trait-based structural match implementation

### LINKED ISSUE #63438: "Structural match" annotations should be moved from struct/enum items to impl items
"Structural match" property of a type means that the constants of this type can be used in patterns (https://github.com/rust-lang/rust/issues/31434) or in const generics in the future.

Right now this property can be given to a type in two ways:
- `#[structural_match]` attribute, rarely used, `PhantomData` is the only example.
- If the struct/enum has both `#[derive(PartialEq)]` and `#[derive(Eq)]`  then it's marked as "structural match". Derives cannot modify the items they are applied to, so this tracking is done through a special table in resolver (which is kind of a hack).

To avoid hacks we need to pass the knowledge about "structural match" not through types themselves, but through impls on them (which *can* be generated by derives easily).

The condition for the type being "structural match" is:
- `PartialEq` methods are *derived* so their behavior is fully known to the compiler.
- `Eq` is *implemented*, but not necessarily derived, it's just a marker without methods.

So, the "structural match" annotation needs to be generated by `derive(PartialEq)` in  the form of either
- an attribute `#[structural_match] impl PartialEq for ... ` (in this case the const check will be "implements `Eq` + implements `PartialEq` + the `PartialEq` impl has the attribute"), or
- an impl of some third trait `impl StructuralMatch for ...` (in this case the const check will be "implements `Eq` + implements `StructuralMatch`").

cc https://github.com/rust-lang/rust/pull/63248 @matthewjasper @varkor @Centril @pnkfelix

### PR DESCRIPTION
**Pull request by pnkfelix**
Moves from using a `#[structural_match]` attribute to using a marker trait (or pair of such traits, really) instead.

Fix #63438.

(This however does not remove the hacks that I believe were put into place to support the previous approach of injecting the attribute based on the presence of both derives... I have left that for follow-on work.)

### CODE REVIEW

**Comment by pnkfelix** [comment]
cc @petrochenkov

**Comment by pnkfelix** [comment]
(I do have other experimental work that goes further in fixing bugs in our structural-match system. However, I decided this piece of the puzzle was worth landing on its own, since it is both broad in the amount of code it touches, but very narrow in its actual user-visible effects...)

File: src/librustc_mir/hair/pattern/const_to_pat.rs [hunk 1]
```diff
@@ -0,0 +1,406 @@
...
+
+        ty.visit_with(&mut search);
+        return search.found;
+    }
+}
+
+struct Search<'a, 'tcx> {
+    const_to_pat: &'a ConstToPat<'a, 'tcx>,
+
+    // records the first ADT we find that does not implement `Structural`.
```
Comment (Centril) [inline src/librustc_mir/hair/pattern/const_to_pat.rs]: Doc comments (+ capitalization) would be better. :)

File: src/librustc_mir/hair/pattern/const_to_pat.rs [hunk 2]
```diff
@@ -0,0 +1,406 @@
...
+    pub(super) fn const_to_pat(
+        &self,
+        cv: &'tcx ty::Const<'tcx>,
+        id: hir::HirId,
+        span: Span,
+    ) -> Pat<'tcx> {
+        debug!("const_to_pat: cv={:#?} id={:?}", cv, id);
+        debug!("const_to_pat: cv.ty={:?} span={:?}", cv.ty, span);
+
+        return self.tcx.infer_ctxt().enter(|infcx| {
```
Comment (Centril) [inline src/librustc_mir/hair/pattern/const_to_pat.rs]:
```suggestion
        self.tcx.infer_ctxt().enter(|infcx| {
```

File: src/librustc_mir/hair/pattern/const_to_pat.rs [hunk 3]
```diff
@@ -0,0 +1,406 @@
...
+    /// Rust RFC 1445, rust-lang/rust#61188, and rust-lang/rust#62307.
+    fn search_for_adt_without_structural_match(&self, ty: Ty<'tcx>) -> Option<&'tcx AdtDef> {
+        let mut search = Search {
+            const_to_pat: self,
+            found: None,
+            seen: FxHashSet::default()
+        };
+
+        ty.visit_with(&mut search);
+        return search.found;
```
Comment (Centril) [inline src/librustc_mir/hair/pattern/const_to_pat.rs]:
```suggestion
        search.found;
```

File: src/librustc_mir/hair/pattern/const_to_pat.rs [hunk 4]
```diff
@@ -0,0 +1,406 @@
...
+}
+
+impl<'a, 'tcx> TypeVisitor<'tcx> for Search<'a, 'tcx> {
+    fn visit_ty(&mut self, ty: Ty<'tcx>) -> bool {
+        debug!("Search visiting ty: {:?}", ty);
+
+        let (adt_def, substs) = match ty.kind {
+            ty::Adt(adt_def, substs) => (adt_def, substs),
+            ty::RawPtr(..) => {
+                // structural-match ignores substructure of
```
Comment (Centril) [inline src/librustc_mir/hair/pattern/const_to_pat.rs]: This case could use some elaboration.

**Comment by Centril** [comment]
cc @varkor

**Comment by pnkfelix** [comment]
> The job x86_64-gnu-llvm-6.0 of your PR failed

(Hmm apparently I need to include a few extra suites in my local test runs.)

**Comment by pnkfelix** [comment]
One thing we probably will want is a perf run on this. I do not know if there might be an impact from the way I implemented `type_marked_structural`.

(If necessary, we could batch up the obligations and try to solve them all at once, but its tricky because we want to have hard errors in some cases and lints in others, due to future-compatibility concerns. That trickiness is why I opted to take this approach first.)

File: src/librustc_mir/hair/pattern/const_to_pat.rs (hunk 1 above)
Comment (pnkfelix) [reply src/librustc_mir/hair/pattern/const_to_pat.rs]: done

File: src/librustc_mir/hair/pattern/const_to_pat.rs (hunk 2 above)
Comment (pnkfelix) [reply src/librustc_mir/hair/pattern/const_to_pat.rs]: done

File: src/librustc_mir/hair/pattern/const_to_pat.rs (hunk 3 above)
Comment (pnkfelix) [reply src/librustc_mir/hair/pattern/const_to_pat.rs]: done

File: src/librustc_mir/hair/pattern/const_to_pat.rs (hunk 4 above)
Comment (pnkfelix) [reply src/librustc_mir/hair/pattern/const_to_pat.rs]: I expanded the comment with an example, hopefully that suffices.

File: src/libsyntax_ext/deriving/mod.rs [hunk 5]
```diff
@@ -74,3 +74,94 @@ fn call_intrinsic(cx: &ExtCtxt<'_>,
...
+            // Non-Item derive is an error, but it should have been
+            // set earlier; see
+            // libsyntax/ext/expand.rs:MacroExpander::expand()
+            return;
+        }
+    };
+
+    let generics = match item.kind {
+        ItemKind::Struct(_, ref generics) => generics,
+        ItemKind::Enum(_, ref generics) => generics,
```
[the remark this answers is not in the document]
Comment (pnkfelix) [reply src/libsyntax_ext/deriving/mod.rs]: done

File: src/librustc_mir/hair/pattern/const_to_pat.rs (hunk 4 above)
Comment (Centril) [reply src/librustc_mir/hair/pattern/const_to_pat.rs]: Yep, it is great, thanks!

File: src/libsyntax_ext/deriving/mod.rs [hunk 6]
```diff
@@ -74,3 +74,94 @@ fn call_intrinsic(cx: &ExtCtxt<'_>,
...
+                                   deriving_sym: Symbol,
+                                   structural_path: generic::ty::Path<'_>,
+                                   push: &mut dyn FnMut(Annotatable)) {
+
+    // set the expn ID so we can use the unstable trait.
+    let span = span.fresh_expansion(syntax_pos::ExpnData::allow_unstable(
+        syntax_pos::ExpnKind::Macro(syntax_pos::MacroKind::Derive, deriving_sym),
+        span,
+        cx.parse_sess.edition,
+        [sym::structural_match][..].into()));
```
Comment (petrochenkov) [inline src/libsyntax_ext/deriving/mod.rs]: `fresh_expansion` shouldn't be used here, instead `structural_match` should go into `#[allow_internal_unstable]` on the `macro PartialEq` and `macro Eq` definitions in libcore.

File: src/librustc_mir/hair/pattern/const_to_pat.rs [hunk 7]
```diff
@@ -0,0 +1,422 @@
...
+    /// type.
+    ///
+    ///  * This means this method descends into fields of structs/enums,
+    ///    and also descends into the inner type `T` of `&T` and `&mut T`
+    ///
+    ///  * The traversal doesn't dereference unsafe pointers (`*const T`,
+    ///    `*mut T`), and it does not visit the type arguments of an
+    ///    instantiated generic like `PhantomData<T>`.
+    ///
+    /// The reason we do this search is Rust currently require all ADT's
```
Comment (varkor) [inline src/librustc_mir/hair/pattern/const_to_pat.rs]:
```suggestion
    /// The reason we do this search is Rust currently require all ADTs
```

File: src/librustc_mir/hair/pattern/const_to_pat.rs [hunk 8]
```diff
@@ -0,0 +1,422 @@
...
+        if self.seen.contains(adt_def) {
+            debug!("Search already seen adt_def: {:?}", adt_def);
+            // let caller continue its search
+            return false;
+        }
+
+        if !self.const_to_pat.type_marked_structural(ty) {
+            debug!("Search found ty: {:?}", ty);
+            self.found = Some(&adt_def);
+            return true // Halt visiting!
```
Comment (varkor) [inline src/librustc_mir/hair/pattern/const_to_pat.rs]:
```suggestion
            return true; // Halt visiting!
```

File: src/libsyntax_ext/deriving/mod.rs (hunk 6 above)
Comment (pnkfelix) [reply src/libsyntax_ext/deriving/mod.rs]: done, thanks!

**Pull request merged**